#7276 Live Preview highlight customization

Worie opened the pull request:
**Live Preview highlight customization (colors of highlight can be customized via user preferences)**

I needed to easly customize the highlight color of Live Preview - and I've achieved it with the changes in the PR. I'm not sure if that's the desired way of doing so, as this is my first PR to Brackets.

1. RemoteFunctions takes whole config as a parameter instead just "experimental" flag.
2. There's a preference in which user can define desired styling of highlightning. 
3. The view is updated when preferences change without breaking live dev session.
4. If no user preferences are defined, the default configuration is used.


It also implements #7276 -> [Preview of Live Preview with margins and paddings showing](http://imgur.com/Gx1r4P3)

Worie committed:
Live Preview visual customization (colors of highlight can be customized via user preferences)

petetnt commented on src/LiveDevelopment/Agents/RemoteAgent.js:
```diff
@@ -128,10 +128,10 @@ define(function RemoteAgent(require, exports, module) {
-        // inject RemoteFunctions
-        var command = "window._LD=" + RemoteFunctions + "(" + LiveDevelopment.config.experimental + ");";
-
+        
```
Nit: trailing whitespace

petetnt commented on src/LiveDevelopment/main.js:
```diff
@@ -79,6 +79,21 @@ define(function main(require, exports, module) {
+    // "livedev.remoteHighlight" preference
+    var PREF_REMOTEHIGHLIGHT = "remoteHighlight";
+    var remoteHighlightPref = prefs.definePreference(PREF_REMOTEHIGHLIGHT, "object", {
+        "animateStartValue": {
```
nit: Keys don't need to be quoted

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -268,16 +269,17 @@ function RemoteFunctions(experimental) {
                 "box-sizing": "border-box"
             };
+            
+            var preferences = config.remoteHighlight.stylesToSet;
```
You can use `Object.assign` here to simplify the merging:

``` js
var mergedStyles = Object.assign({}, stylesToSet,  config.remoteHighlight.stylesToSet)
```

and then later use `mergedStyles` instead of `stylesToSet`

petetnt commented on src/LiveDevelopment/Agents/RemoteAgent.js:
```diff
@@ -128,10 +128,10 @@ define(function RemoteAgent(require, exports, module) {
+        
+        // inject RemoteFunction
+        var command = "window._LD=" + RemoteFunctions + "(" + JSON.stringify(LiveDevelopment.config) + ");";
+        
```
Nit: trailing whitespace

petetnt reviewed (changes requested):
Nice work @Worie! Added some small comments.

Might require a test too.

Worie committed:
PR improvements
Remove trailing spaces, use Object.assign instead of manually iterating over remoteHighlight properties and remove quotes from object properties in default configuration of remoteHighlight

Worie commented:
Thanks! I've applied the changes you have mentioned. Unfortunetly it seems like I've broken the test, as `Debug -> Run tests` throws `Uncaught TypeError: Cannot read property 'experimental' of undefined` in `SpecRunner.html:36` . I can't investigate on it now, though.

Also it looks like I've closed this PR 3 hours ago, is it automated process or did i just close it by accident?

petetnt commented:
These closing is not an automated process, I guess you closed it by accident but that's no biggie 👍 

I can help you out with debugging the test failures in a while so we can land this 👍

Worie committed:
Allow to animate transform property via transition

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -268,21 +269,17 @@ function RemoteFunctions(experimental) {
 
             var transitionValues = {
-                "-webkit-transition-property": "opacity, background-color",
+                "-webkit-transition-property": "opacity, background-color, transform",
```
I think we don't need to support the prefixed ones anymore

Worie committed:
Remove old vendor prefixes from transitionValues in RemoteFunctions.js

Worie commented:
It isn't pushed yet, but I've created `updateConfig` function in `RemoteFunctions.js`  so the changes in the `brackets.json` could be loaded to current LiveDev session. 

Also I'd like to try to implement https://github.com/adobe/brackets/issues/7276 in my free time, as I think it is an awesome case  and it could play well with changes made in this PR.

Worie committed:
updateConfig RemoteFunctions.js method
Enable injecting new config into remoteFunction when user changed preferences, without the need to restart the live preview session.

ficristo commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -268,21 +269,15 @@ function RemoteFunctions(experimental) {
                 "box-sizing": "border-box"
             };
+            
+            var mergedStyles = Object.assign({}, stylesToSet,  config.remoteHighlight.stylesToSet);
```
I think Object.assign is not available on Linux (which uses Chrome 29). Could you try lodash assign or extend instead.

Worie replied:
Those are remote functions, so I cannot use `_.assign` there I guess until i inject it to the browser. Am I right? 

I've left it as `Object.assign`, just for now - until I know what exactly am I supposed to do.

petetnt replied:
Actually as they are remote functions that are run on the browser, we can just use `Object.assign` right?

ficristo replied:
You mean a real browser and not Brackets itself, right? (Not much familiar with LivePreview)
Then I suppose we can use `Object.assign`. Sorry for the noise...

Worie replied:
I cant test it on Linux, but it works well on OSX

Worie committed:
Show margin and paddings in live preview.

Worie commented:
[Preview of Live Preview with margins and paddings showing](http://imgur.com/LDjNmEQ)

So, I've done it :) If user did not specify his own preferences, it looks as designed in https://github.com/adobe/brackets/issues/7276 . Also changing those values does not break current Live Development session.

I could probably add one condition to check if there's any padding, so there wont be any borders showing when padding value is equal to 0 or something.

Worie committed:
Do not show element if it has width or height equal to 0. A little bit of clean up

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -244,6 +245,173 @@ function RemoteFunctions(experimental) {
+            var calculateSize = function () {
+                var sum = 0;
+                var i;
+                for (i = 0; i < arguments.length; i++) {
```
Not a huge one of using `arguments`, can we instead pass an array and loop through that (with `Array.prototype.reduce` for example?)

```
var calculateSize = function (values) {
  return values.reduce(function (val) {
    return parseInt(val, 10);
  }, 0);
};
```

(Side note, I wish we could use ES2015 so this would be just `const calculateSize = (values) => values.reduce(val => parseInt(val, 10), 0)` 😸)

petetnt replied:
`calculateSize` should probably return value in pixels too instead of using string concat on every use.

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -244,6 +245,173 @@ function RemoteFunctions(experimental) {
+            
+            var mainBoxStyles = config.remoteHighlight.stylesToSet;
+            
+            var paddingVisualisations = [
```
Not sure if these would be better in their own file, for example `RemoteVisualizationUtils`  for example?

Worie replied:
Hmm, but `RemoteFunctions` are being injected into the browser (`window._LD`) and those values are calculated basing on the element that is highlighted there... I'm not sure how that sould work - how to access another file from there? Can't imagine this atm.

Or perhaps i simply didn't understand what you meant :)

petetnt replied:
Yeah my bad, I was thinking about making it an utility file of methods that get passed to that injected method but it most likely won't work like that nor would be worth it. Feel free to disregard 👍

Worie committed:
PR improvements
calculateSize is renamed to sum. (better naming conventions are welcome). sum now takes an array as argument. It returns sum of parsed array elements, suffixed with „px”.

Worie commented:
Realized that elements that have css `transition` on them act a bit buggy (margins and paddings do not animate, and it causes that you have to wait for the transition to finish, then re-select this specific node to see changes), perhaps I'll do something about it next week, we'll see. 

But still, i have no idea how to get into fixing those tests - any help would be awesome, even pointing to some specific files or anything. :)

ficristo commented:
If you open Chrome at http://localhost:9234/, after clicked `Debug \ Run Tests`, there are two errors `Uncaught TypeError: Cannot read property 'experimental' of undefined`.
Their stack trace point to:
[spec/HTMLInstrumentation-test.js:40](https://github.com/adobe/brackets/blob/75406e8c94062a7cd06204632eda78494b8732ef/test/spec/HTMLInstrumentation-test.js#L40)
[spec/RemoteFunctions-test.js:33](https://github.com/adobe/brackets/blob/75406e8c94062a7cd06204632eda78494b8732ef/test/spec/RemoteFunctions-test.js#L33)

Worie committed:
Redraw via requestAnimationFrame if element has transition / animation on it.

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -30,10 +30,24 @@
 
+    var experimental = config.experimental;
     var lastKeepAliveTime = Date.now();
+	var req, timeout;
```
This commit has `tabs`, while we prefer/enforce 4 spaces. You can easily change them with the `brackets-eslint` extension by right clicking the editor and selecting `Autofix with ESLint`

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -233,17 +247,190 @@ function RemoteFunctions(experimental) {
+			
+		var animationDuration = parseFloat(styles.getPropertyValue('animation-duration'));
+			
+		if (transitionDuration)
```
Nit: IIRC we prefer to have braces on single line ifs too:

```
if (transitionDuration) {
  animateHighlight(transitionDuration);
}
if (animationDuration) {
    animateHighlight(animationDuration);
}

Worie replied:
Yeah, I do too. I have no idea why did i write it like so. Thanks!

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -233,17 +247,190 @@ function RemoteFunctions(experimental) {
             var elementBounds = element.getBoundingClientRect(),
                 highlight = window.document.createElement("div"),
                 styles = window.getComputedStyle(element);
+		var transitionDuration = parseFloat(styles.getPropertyValue('transition-duration'));
```
Nit: these two variables could also be combined with the above by separating the values with , like the others are.

petetnt reviewed (changes requested):
Few more small comments. Nice work so far 👍

Worie committed:
PR improvements.
Set experimental to false by default, if no config specified (caused the test to crash).
Minor code quality improvements, removed duplicated variables.

Worie committed:
Fix ESLINT errors.

Worie commented:
Due to `getClientBoundingRect` including the transform when calculating the size of element and its position (https://i.imgur.com/tNNjXEw.png), i get pretty weird and unintuitive result - http://imgur.com/MTwZcfB .

I do not know how to bite this one without some heavy math calculations (nor I am any good at math).

Perhaps some of you have an idea how to get the current implementation to act like a chrome one easly?

petetnt commented:
@Worie I was thinking about that but didn't find any simple solutions yet, could you share an sample project / file that shows the behaviour in the GIF?

Worie commented:
Perhaps this will be enough: https://gist.github.com/Worie/caae957b805f08cb733f51406d32aabe 

Also, now every transform function is applied to the highlight div - probably only scale and rotate should be considered.

Worie commented:
I've started a little refactor of this code. I'll not focus on the transform issue atm, though.

Worie committed:
drawMarginRect and drawPaddingRect functions now calculate the require size of highlight elements instead of manaul calculations on each element.

The code is shorter and more readable. Also, it’ll be easier to move drawing functions into separate file if necessary.

Worie commented:
@petetnt In last commit, I've removed the `sum` function we introduced during this PR. I felt that it was making more harm than good and made it less clear what exacly is happening there. 

Of course, I can revert this change and apply the `sum` function to current implementation. But I guess it's not really necessary.

petetnt commented:
Hey Worie,

sorry for the lack of comms, been a bit busy as of late. Is this PR ready for another round of review?

Worie commented:
Yeah I guess. If we could only _somehow_ invoke the `inspect` on element as we can in Chrome Dev tools from `RemoteFunctions` this PR would be much shorter :/ But I don't think it is possible atm.

Anyway - the workaround for transform issue is that we simply do not apply transform to our custom highlightning divs. That means that they dont rotate nor scale, but translation works pretty good on them. 

I guess that we could perhaps split this into two issues - close this one when the code quality is good enough, and then take care of the transform issue. 

Because, I think that it might take a while to do (and I have no idea atm how, but I guess we'd need to get the data about the transformations from matrix3d and reverse it to something like `rotate(45deg) translateX()...`)

Worie committed:
Temporary workaround for transform issue.

petetnt commented:
I think the scope is just fine right now, I could run another code review and test it out again myself and then merge it in (and create a issue for tracking the transform part). @ficristo you could also take a look if you want to 👍

ficristo commented:
I think @saurabh95 is more indicated

CodingAnonymous commented:
When will the Live Preview Highlight with Margin and Padding be available? Or can I activate it manually now?
Thanks for your help!

CodingAnonymous commented:
This feature is so great and important, can I use it right now through some code implementation? http://imgur.com/Gx1r4P3

Worie commented:
You can always clone my fork and set it as a source for Brackets, see [how to set up dev environment](https://github.com/adobe/brackets/wiki/How-to-Hack-on-Brackets#setting-up-your-dev-environment).

Can I somehow help you guys with this PR? I mean, we could discuss the code that's already here, or perhaps I can tell you my way of thinking when i was writing this. Be sure to ping me when anything like this will be necessary.

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -30,10 +30,29 @@
+    }
     var lastKeepAliveTime = Date.now();
+	var req, timeout;
+	var animateHighlight = function (time) {
```
L44-L55 have some indentation issues, convert tabs -> spaces and they should be all right

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
             }
+            
+            var realElBorder = {
+              "right": elementStyling.getPropertyValue('border-right-width'),
```
No need to quote the keys in L293-L296

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
+            var innerWidth = elementBounds.width - parseInt(realElBorder.left) - parseInt(realElBorder.right);
+            var innerHeight = elementBounds.height - parseInt(realElBorder.top) - parseInt(realElBorder.bottom);
+          
+            var visualisations = [];
```
This should be an object literal instead of assigning `horizontal` and `vertical` as array properties?

``` js
var visualisations = {
 horizontal: "left, right",
 vertical: "top, bottom",
}
```

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
+            var drawPaddingRect = function(side) {
+              var elStyling = {};
+              
+              if(visualisations['horizontal'].indexOf(side) >= 0) {
```
Maybe prefer the dot-notation (`visualisations.horizontal` etc.) in L309-351.

petetnt replied:
Nit: space between `if` and `(` -> `if (visualisations)...`

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
+          var drawMarginRect = function(side) {
+            var elStyling = {};
+            
+            var  margin = [];
```
Nit: extra space between `var` and `margin`

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
+            };
+            
+            var mainBoxStyles = config.remoteHighlight.stylesToSet;
+                mainBoxStyles['border'] = 'none';
```
Nit: `mainBoxStyles` should be aligned to `var`

petetnt commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -250,17 +269,149 @@ function RemoteFunctions(experimental, remoteWSPort) {
+                        arr[i]
+                    );
+
+                    _setStyleValues(
```
Nit: This should be a one-liner as there's only couple of short properties

Worie replied:
Oh, okay. It's common practice for me to place parameters (even short ones) in separate lines so git displays the changes in more friendly way. But sure, I'll make it a one-liner. 👍

petetnt reviewed (changes requested):
Sorry about taking so long @Worie, this got a bit buried between other work.

The implementation looks a bit complex, but then again what it achieves is also a bit complex so that is okay.

I put some small review comments about the code style, and one that is most likely an error, but otherwise it looks pretty good to me.

You can run the files through `eslint` to catch some other errors there might be if any. 👍

Commit:
Formatting improvements

saurabh95 commented on src/LiveDevelopment/Agents/RemoteFunctions.js:
```diff
@@ -30,10 +30,29 @@
  * exported functions.
  */
-function RemoteFunctions(experimental, remoteWSPort) {
+function RemoteFunctions(config, remoteWSPort) {
```
I think we can pass remoteWSPort as part of config only, something like `config.remoteWSPort`.
Then instead of checking for `remoteWSPort` in https://github.com/adobe/brackets/pull/12949/files#diff-a40df952af8fd1808c321dab548ae11aR1032, we can check for `!experimental` flag as we create websocket only when we are not using experimental live preview.
But I think you can leave this for later as you might have to make changes which are not relevant for this feature, once this is merged I can make the above changes.

Worie replied:
Hi, you're right. Unfortunetely I won't be accessible for a while, so I guess we could close this one and open new PRs as necessary, if that's OK.

Worie commented:
@petetnt what do you think?

petetnt commented:
Also thanks for this great contribution and sorry that it took so long to get merged 👍

Worie commented:
Awesome! Sure, I'll create issues that were not covered in this PR. Thanks!
